core/arch/arm/mm/mobj.c

```diff
@@ -0,0 +1,432 @@
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <assert.h>
+#include <kernel/mutex.h>
+#include <kernel/panic.h>
+#include <kernel/tee_misc.h>
+#include <mm/core_mmu.h>
+#include <mm/mobj.h>
+#include <mm/tee_mmu.h>
+#include <mm/tee_pager.h>
+#include <optee_msg.h>
+#include <sm/optee_smc.h>
+#include <stdlib.h>
+#include <tee_api_types.h>
+#include <types_ext.h>
+#include <util.h>
+
+struct mobj *mobj_sec_ddr;
+
+/*
+ * mobj_phys implementation
+ */
+
+struct mobj_phys {
+	struct mobj mobj;
+	enum buf_is_attr battr;
+	uint32_t cattr; /* Defined by TEE_MATTR_CACHE_* in tee_mmu_types.h */
+	vaddr_t va;
+	paddr_t pa;
+};
+
+static struct mobj_phys *to_mobj_phys(struct mobj *mobj);
+
+static void *mobj_phys_get_va(struct mobj *mobj, size_t offset)
+{
+	struct mobj_phys *moph = to_mobj_phys(mobj);
+
+	if (!moph->va)
+		return NULL;
+
+	return (void *)(moph->va + offset);
+}
+
+static TEE_Result mobj_phys_get_pa(struct mobj *mobj, size_t offs,
+				   size_t granule, paddr_t *pa)
+{
+	struct mobj_phys *moph = to_mobj_phys(mobj);
+	paddr_t p;
+
+	if (!pa)
+		return TEE_ERROR_GENERIC;
+
+	p = moph->pa + offs;
+
+	if (granule) {
+		if (granule != SMALL_PAGE_SIZE &&
+		    granule != CORE_MMU_PGDIR_SIZE)
+			return TEE_ERROR_GENERIC;
+		p &= ~(granule - 1);
+	}
+
+	*pa = p;
+	return TEE_SUCCESS;
+}
+
+static TEE_Result mobj_phys_get_cattr(struct mobj *mobj, uint32_t *cattr)
+{
+	struct mobj_phys *moph = to_mobj_phys(mobj);
+
+	if (!cattr)
+		return TEE_ERROR_GENERIC;
+
+	*cattr = moph->cattr;
+	return TEE_SUCCESS;
+}
+
+static bool mobj_phys_matches(struct mobj *mobj, enum buf_is_attr attr)
+{
+	struct mobj_phys *moph = to_mobj_phys(mobj);
+	enum buf_is_attr a;
+
+	a = moph->battr;
+
+	switch (attr) {
+	case CORE_MEM_SEC:
+		return a == CORE_MEM_SEC || a == CORE_MEM_TEE_RAM ||
+		       a == CORE_MEM_TA_RAM;
+	case CORE_MEM_NON_SEC:
+		return a == CORE_MEM_NSEC_SHM;
+	case CORE_MEM_TEE_RAM:
+	case CORE_MEM_TA_RAM:
+	case CORE_MEM_NSEC_SHM:
+		return attr == a;
+	case CORE_MEM_MULTPURPOSE:
+		return a == CORE_MEM_TA_RAM || CORE_MEM_NSEC_SHM;
```

Comment:
typo: `|| a == CORE_MEM_NSEC_SHM`

Old MEM_MULTIPURPOSE  was based on checking is physical address range for mobj lies inside DDR and outside TZ "Secure" reserved memory (TEE_RAM, TA_RAM and NS_SHM).